plutus-ledger-api/src/PlutusLedgerApi/Common.hs

```diff
@@ -1,73 +1,138 @@
+
+  -- * Script evaluation
+  Eval.evaluateScriptCounting,
+  Eval.evaluateScriptRestricting,
+  Eval.evaluateTerm,
+  Eval.VerboseMode (..),
+  Eval.LogOutput,
+  Eval.EvaluationError (..),
+
+  -- * Network's versioning
+  {-| The network's behaviour (and plutus's by extension) can change via /hard forks/,
+  which directly correspond to major-number protocol version bumps.
+  -}
+  Versions.MajorProtocolVersion (..),
+  Versions.PlutusLedgerLanguage (..),
+  Versions.Version (..),
+  Versions.builtinsIntroducedIn,
+  Versions.builtinsAvailableIn,
+  Versions.ledgerLanguageIntroducedIn,
+  Versions.ledgerLanguagesAvailableIn,
+
+  -- * Costing-related types
+  PLC.ExBudget (..),
+  PLC.ExCPU (..),
+  PLC.ExMemory (..),
+  SatInt.SatInt (unSatInt),
+  SatInt.fromSatInt,
+
+  -- * Network's costing parameters
+  {-| A less drastic approach (that does not rely on a HF)
+  to affect the network's (and plutus's by extension) behaviour
+  is by tweaking the values of the cost model parameters.
+
+  The network does not associate names to cost model parameters;
+  Plutus attaches names to the network's cost model parameters (values)
+  either in a raw textual form or typed by a specific plutus version.
+
+  See Note [Cost model parameters]
+  -}
+  PLC.CostModelParams,
+  ParamName.toCostModelParams,
+  Eval.assertWellFormedCostModelParams,
+  ParamName.IsParamName (showParamName, readParamName),
+  ParamName.GenericParamName,
+  ParamName.CostModelApplyError (..),
+  ParamName.CostModelApplyWarn (..),
+
+  -- ** Evaluation context
+  Eval.EvaluationContext (..),
+  Eval.mkDynEvaluationContext,
+  Eval.toMachineParameters,
+  -- While not strictly used by the ledger, this is useful for people trying to
+  -- reconstruct the term evaluated by the ledger from the arguments, e.g.
+  -- for profiling purposes.
+  Eval.mkTermToEvaluate,
+
+  -- ** Supporting types used in the context types
+
+  -- *** Builtins
+  Prelude.BuiltinByteString,
+  Prelude.toBuiltin,
+  Prelude.fromBuiltin,
+  Prelude.toOpaque,
+  Prelude.fromOpaque,
+
+  -- * Data
+  PLC.Data (..),
+  Builtins.BuiltinData (..),
+  IsData.ToData (..),
+  IsData.FromData (..),
+  IsData.UnsafeFromData (..),
+  IsData.toData,
+  IsData.fromData,
+  IsData.unsafeFromData,
+  Builtins.dataToBuiltinData,
+  Builtins.builtinDataToData,
+
+  -- * Misc
+  MonadError,
+) where
+
+import PlutusLedgerApi.Common.Eval qualified as Eval (EvaluationContext (..), EvaluationError (..),
+                                                      LogOutput, VerboseMode (..),
+                                                      assertWellFormedCostModelParams,
+                                                      evaluateScriptCounting,
+                                                      evaluateScriptRestricting, evaluateTerm,
+                                                      mkDynEvaluationContext, mkTermToEvaluate,
+                                                      toMachineParameters)
```

Comment:
Made import lists explicit in this module. I normally don't care much about readability of imports, but given that it's one of the ledger API modules and it's important to keep it reasonable and readable, I thought it might be worth going full verbosity. Now it always clear where a definition comes from, regardless of whether you look at imports (due to the explicit list) or exports (due to everything being imported `qualified` apart from entirely trivial things, of which we only have `MonadError`).

Please tell me if I should remove the explicit imports or add them to other modules or keep everything the same.

Reply:
> I thought it might be worth going full verbosity

I believe this can be enforced by HLint, so while you are at it you may want to add an HLint rule to ensure it stays this way.

> Please tell me if I should remove the explicit imports or add them to other modules or keep everything the same.

In general having _both_ qualified imports _and_ explicit imports is an overkill, so I wouldn't do this to other modules - I don't see the need. I'm in favor of using _either_ qualified imports _or_ explicit imports everywhere, but others even find this to be not worth it.

Reply:
My take:
- Wildcard imports: mostly no, could be exceptions.
- Qualified imports: nice, should use more.
- Explicit imports: also nice, complements qualified:

```haskell
import Data.Text qualified as Text

message :: Text.Text -- IMO this duplication is ugly
message = Text.toUpper "hello" -- this is readable.
```

the combo of qualified and explicit is what I'd use:

```haskell
import Data.Text qualified as Text
import Data.Text (Text)

message :: Text -- The best 
message = Text.toUpper "hello" -- good
```

Reply:
> I believe this can be enforced by HLint, so while you are at it you may want to add an HLint rule to ensure it stays this way.

Sorry, I'm fully committed to never adding HLint anywhere. I was forced to use it for months and it damaged me badly.

> In general having both qualified imports and explicit imports is an overkill

I totally agree, it's just that these modules are very special in that everything exported by them must be scrutinized deeply. My usual standards for imports are lower than yours or Yura's.

@Unisay I agree with you, I'm just usually lazy when it comes to import management. Especially if it's within "standard" modules like `base` or `PlutusCore` or whatever. I'll try to do a better job.